## BUSA8001- Programming Task 1  

**Assignment Points**: 100  
**Submission**: Provide your answers in this notebook and submit it via iLearn

- Where a question requires a written answer provide your solution in Markdown in the cells under each question.
- Comment out your print statements unless you are explicitly asked to print your output. 
- 5 marks will be deducted for printed outputs that are not asked for.

### About the Assignment

- For this assignment there are two files in the `data` folder `credit_record.csv` and `application_record.csv` where bank clients are related by the `ID` column.

- In `application_record.csv` we have the following variables

| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number      |         |
| AMT_INCOME   | Annual income  |  |
| NAME_INCOME_TYPE   | Income Source |  |
| NAME_EDUCATION_TYPE   | Level of Education  |  |
| CODE_GENDER   | Applicant's Gender   |  |
| FLAG_OWN_CAR | Car Ownership |  | 
| CNT_CHILDREN | Number of Children | |
| FLAG_OWN_REALTY | Real Estate Ownership | | 
| NAME_FAMILY_STATUS | Relationship Status | | 
| NAME_HOUSING_TYPE | Housing Type | | 
| DAYS_BIRTH | No. of Days | Count backwards from current day (0), -1 means yesterday
| DAYS_EMPLOYED | No. of Days | Count backwards from current day (0). If positive, it means the person is currently unemployed.
| FLAG_MOBIL | Mobile Phone Ownership | | 
| FLAG_WORK_PHONE | Work Phone Ownership | | 
| FLAG_PHONE | Landline Phone Ownership | | 
| FLAG_EMAIL | Landline Phone Ownership | | 
| OCCUPATION_TYPE | Occupation | | 
| CNT_FAM_MEMBERS | Count of Family Members | |



- In `credit_record.csv` we have the following variables


| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number | |
| MONTHS_BALANCE | Number of months in the past from now when STATUS is measured | 0 = current month, -1 = last month, -2 = two months ago, etc.|
| STATUS | Number of days a payment is past due | 0: 1-29 days past due 1: 30-59 days past due 2: 60-89 days overdue 3: 90-119 days overdue 4: 120-149 days overdue 5: Overdue or bad debts, write-offs for more than 150 days C: paid off that month X: No loan for the month |

---
---

### Task 1: Reading, Summarising and Cleaning Data (Total Marks: 30)

**Question 1.** 

1. Import the `application_record.csv` and `credit_record.csv` files from `data` folder into pandas DataFrames named `df_application` and `df_credit`, respectively. (1 mark)

2. How many rows are there in `df_application` and `df_credit`, respectively? Provide your answers with print() and state them in Markdown text. (1 mark)

3. How many unique bank clients are there in `df_application` and `df_credit`? Provide your answers with print() and state them in Markdown text. (1 mark)

4. Add the records from `df_credit` to `df_application` by merging the data from the two DataFrames on the `ID` column, and output the joint data into a new DataFrame named `df`. Hint: Use `merge` function from pandas by setting `how` parameter to `inner` (4 marks) 

5. How many rows and how many unique clients are there now in `df`? (1 mark)

6. How are multiple rows for each `ID` in `df` different? Answer in Markdown text. (2 mark) 

(10 marks)

In [37]:
# ---- provide your code here -----
# This code file has to be stored at the same folder and same folder level as the datas
# Question 1.1

import pandas as pd #import pandas lib

df_application=pd.read_csv('application_record.csv') #import application record 

df_credit=pd.read_csv('credit_record.csv')  #import credit record

#Question 1.2

#count column for df_application
print("Number of rows in df_application is ", len(df_application))

#count column for df_credit
print("Number of rows in df_credit is ", len(df_credit))

#Question 1.3

#count number of unique bank clients for df_application
print("Number of unique bank clients in df_application is ", len(df_application['ID'].unique()))
#count number of unique bank clients for df_credit
print("Number of unique bank clients in df_credit is ", len(df_credit['ID'].unique()))

#Question 1.4

df=pd.merge(df_application,df_credit, how= 'inner')

#Question 1.5

print(f"df has {len(df)} rows") #count number of rows
print(f"There are {len(df['ID'].unique())} unique clients now in df")   #count number of unique clients


Number of rows in df_application is  438445
Number of rows in df_credit is  1047185
Number of unique bank clients in df_application is  438398
Number of unique bank clients in df_credit is  45924
df has 776325 rows
there are 36396 unique clients now in df


---- provide your written answer here ----

2.
Number of rows in df_application is  438445

Number of rows in df_credit is  1047185

3. 
Number of unique bank clients in df_application is  438398
Number of unique bank clients in df_credit is  45924

5. 
'df' has 776325 rows

There are 36396 unique clients now in 'df'

6. 
According to the display of 'df' data, we can see that for each unique ID, there can be multiple rows with similar data in the other column headings except for 'MONTHS_BALANCE' column. 

This can be explained as the same customer may deposit or withdraw money from the bank multiple time during a period, resulting in the fluctuation of their bank balance



---

**Question 2.**

1. Change the values of `STATUS` in `df` according to the following mapping: {C, X, 0} -> 0 and {1, 2, 3, 4, 5} -> 1 making sure that the new values of 0 and 1 are encoded as integers. (2 marks)

2. Create a new *numpy* array called `list_of_past_due` that includes the unique ID numbers of clients whose `STATUS = 1` at any point during the last 12 months (hint: count the current month as the first month). (2 marks) 

3. Create a new DataFrame called `df_final` that contains the rows of `df` for which the `ID` is in `list_of_past_due`, keeping only one row for each `ID` (hint: keep the first duplicate row). How many rows do you have in `df_final`? Answer using both print() function and in Markdown text. (hint: find out about `isin()` function in pandas.) (2 marks)

4. Add a new column `y = 1` for all the rows in `df_final`. (1 marks)

5. Increase `df_final` to a total of 4,500 rows by adding rows from `df` with unique `ID`s which are not in `list_of_past_due`. To do this start adding the rows from the beginning of `df`. (hint: learn what `~`, i.e. tilde sign, does in pandas). (2 marks) 

6. Fill the missing values of `y` in `df_final` with zeros. Remove `STATUS` and `MONTHS_BALANCE` from `df_final`. (1 mark)

(10 marks)

In [38]:
# ---- provide your code here -----

#Question 2.1

#Method 1: Using replace
df['STATUS'].replace(to_replace={"C","X",0},value=0, inplace=True)
df['STATUS'].replace(to_replace=list(range(1,6)),value=1, inplace=True)

#Medthod 2: Using map
status_mapping={'C':0,'X':0, '0':0, '1':1,'2':1,'3':1,'4':1,'5':1}
df['STATUS']=df['STATUS'].map(status_mapping)


#Question 2.2

import numpy as np
list_of_past_due = df['ID'][(df['STATUS'] == 1) & (df['MONTHS_BALANCE'] > -12)].unique()

#make sure list_of_past_due is numpy array
if isinstance(list_of_past_due, np.ndarray):
    pass
else:
    list_of_past_due=list_of_past_due.to_numpy()
    
#print list_of_past_due
print(list_of_past_due)

#QUestion 2.3

df_final = df[df['ID'].isin(list_of_past_due)].drop_duplicates(subset='ID', keep='first')
print(len(df_final))

#Question 2.4

df_final['y'] = 1

#Question 2.5

#creating a subarray that contains the rows of `df` for which the `ID` is NOT in `list_of_past_due`, keeping only one row for each `ID`
df_sub = df[~df['ID'].isin(list_of_past_due)].drop_duplicates(subset='ID', keep='first')
df_sub=df_sub.head(4500-len(df_final))
df_final=pd.concat([df_final,df_sub],ignore_index=True)


#Question 2.6

df_final.drop(["STATUS","MONTHS_BALANCE"], axis='columns', inplace=True)  #remove STATUS AND MONTHS_BALANCE FROM DF_FINALS
df_final['y'].replace(to_replace=np.nan,value=0, inplace=True)

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_13288\373365566.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['STATUS'].replace(to_replace={"C","X",0},value=0, inplace=True)


[5008831 5008872 5008912 ... 5149834 5150049 5150337]
1737


C:\Users\LENOVO\AppData\Local\Temp\ipykernel_13288\373365566.py:48: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_final['y'].replace(to_replace=np.nan,value=0, inplace=True)


---- provide your written answer here ----

3.3 

According to the result of print(len(df_final)), which outputs the length (or in this case, number of rows) of df_final variable, there are 1737 rows in the df_final variable

<hr style="width:25%;margin-left:0;"> 

**Question 3**. 
1. Delete `ID` column from `df_final` and reset its index. (1 marks)
2. Assuming that `NAME_EDUCATION_TYPE` is the only ordinal variable in `df_final`, which variables are numeric and which ones are nominal? Answer this question by copying and completing the following table (6 marks)

|Variable type|Numbers of features|Features' list|
| --- | --- | --- |
|Numeric:|||
|Ordinal:|1| NAME_EDUCATION_TYPE |
|Nominal:|||

3. Using appropriate functions find and comment on the missing values in `df_final` (3 marks)   

(10 marks)

In [39]:
# ---- provide your code here ----
#3.1
df_final.drop("ID", axis='columns', inplace=True)
df_final.reset_index(drop= True)

#3.3
df_final.isnull() #print data rows containing NULL values
print('Missing Values: \n',df_final.isnull().sum().sort_values()) #find number of Null values for each cloumn

Missing Values: 
 CODE_GENDER               0
FLAG_OWN_CAR              0
FLAG_OWN_REALTY           0
AMT_INCOME                0
NAME_FAMILY_STATUS        0
NAME_INCOME_TYPE          0
DAYS_BIRTH                0
NAME_HOUSING_TYPE         0
FLAG_WORK_PHONE           0
FLAG_PHONE                0
DAYS_EMPLOYED             0
FLAG_MOBIL                0
FLAG_EMAIL                0
CNT_FAM_MEMBERS           0
y                         0
CNT_CHILDREN             74
OCCUPATION_TYPE        1354
NAME_EDUCATION_TYPE    1831
dtype: int64


---- provide your written answer here ----

3.2

|Variable type|Numbers of features|Features' list|
| --- | --- | --- |
|Numeric:|5| CNT_CHILDREN, AMT_INCOME, DAYS_BIRTH, DAYS_EMPLOYED, CNT_FAM_MEMBERS |
|Ordinal:|1| NAME_EDUCATION_TYPE |
|Nominal:|12| CODE_GENDER, FLAG_OWN_CAR, FLAG_OWN_REALITY, NAME_INCOME_TYPE, NAME_FAMILY_STATUS, NAME_HOUSING_TYPE, OCCUPATION_TYPE, FLAG_MOBIL, FLAG_WORK_PHONE, FLAG_PHONE, FLAG_EMAIL, y|

3.3

The missing values in df_final are in three columns (listed in increasing order of missing values for each column): CNT_CHILDREN (74 missing values), OCCUPATION_TYPE (1,354 missing values), and NAME_EDUCATION_TYPE (1,831 missing values). To fill in these null values, we need to encode the data in these columns (if necessary) and replace the blank cells with the following:

CNT_CHILDREN: Use the median of the existing values. Since the number of children must be numeric and a positive integer, choosing the median ensures that these criteria are met.

OCCUPATION_TYPE: Use the mode or the most frequent value. As the type of occupation is nominal, the mode is the most suitable option because the encoded values do not need to represent any order among the listed jobs.

NAME_EDUCATION_TYPE: Use the mode or the most frequent value. Although the type of education is ordinal, meaning it can be arranged in a specific order, the mode is preferable in this case. Using the median would require initial encoding of the data, which will be addressed in Question 5—one question after the data replacement question. Since mode does not require this additional step, it is the more practical choice.


---
---

### Task 2: Imputing missing values and dealing with categorical features (Total Marks: 30)



**Question 4.** 
1. Use an appropriate `pandas` function to impute missing values in `df_final` (15 marks)
    - Take into consideration the type of each variable and the best practices we discussed in class/lecture notes

(Total: 15 marks)

In [40]:
# ---- provide your code here -----
#replace NULL valies using fillna
df_final['CNT_CHILDREN'].fillna(df_final['CNT_CHILDREN'].median(axis=0),inplace=True)
df_final['NAME_EDUCATION_TYPE'].fillna(df_final['NAME_EDUCATION_TYPE'].mode()[0],inplace=True)
df_final['OCCUPATION_TYPE'].fillna(df_final['OCCUPATION_TYPE'].mode()[0],inplace=True)

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_13288\1406477015.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_final['CNT_CHILDREN'].fillna(df_final['CNT_CHILDREN'].median(axis=0),inplace=True)
C:\Users\LENOVO\AppData\Local\Temp\ipykernel_13288\1406477015.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting va

<hr style="width:25%;margin-left:0;"> 

**Question 5**. Convert the values in `NAME_EDUCATION_TYPE` as follows
- Lower secondary -> 1
- Secondary / secondary special -> 2
- Incomplete higher -> 3
- Higher education -> 4


(Total: 5 marks)  

In [41]:
# ---- provide your code here -----

df_final['NAME_EDUCATION_TYPE'].replace(to_replace="Lower secondary",value=int(1), inplace=True)
df_final['NAME_EDUCATION_TYPE'].replace(to_replace="Secondary / secondary special",value=int(2), inplace=True)
df_final['NAME_EDUCATION_TYPE'].replace(to_replace="Incomplete higher",value=int(3), inplace=True)
df_final['NAME_EDUCATION_TYPE'].replace(to_replace="Higher education",value=int(4), inplace=True)

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_13288\1481117792.py:6: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_final['NAME_EDUCATION_TYPE'].replace(to_replace="Higher education",value=int(4), inplace=True)


<hr style="width:25%;margin-left:0;"> 

**Question 6**. 

Add dummy variables to `df_final` for all of the nominal features which are currently stored as string (text). 
- Make sure to delete the original variables from the dataframe
- Drop the first column from each set of created dummy variable, i.e. for each feature



(Total: 10 marks)  

In [42]:
# ---- provide your code here -
list_of_heading=df_final.columns
for i in list_of_heading:
    if df_final[i].dtype=='object':
        one_hot = pd.get_dummies(df_final[[i]], dtype = int, drop_first = True)
        df_final.drop(i,axis='columns',inplace=True)
        df_final=df_final.join(one_hot)

In [2]:
import pandas as pd
diff = pd.DataFrame({'R': ['a', 'c', 'd'], 
                     'T': ['d', 'a', 'c'],
                     'S_': [1, 2, 3]})
 
print(pd.get_dummies(diff, prefix=['column1', 'column2']))

   S_  column1_a  column1_c  column1_d  column2_a  column2_c  column2_d
0   1       True      False      False      False      False       True
1   2      False       True      False       True      False      False
2   3      False      False       True      False       True      False


---
---

### Task 3 Preparing X and y arrays (Total Marks: 10)

**Question 7**. 

1. Create a numpy array named `y` from the `y` column of `df_final` making sure that the values of the array `y` are stored as integers (3 marks)   
2. Create a numpy array named `X`  from all the remaining features in `df_final` (2 marks)   

(Total: 5 Marks)

In [43]:
# ---- provide your code here -----

#7.1
y = df_final['y'].astype(int).values

#7.2
X = df_final.drop(columns=['y'], inplace=False).values

<hr style="width:25%;margin-left:0;"> 

**Question 8**. 

1. Use an appropriate scikit-learn library we used in class to create `y_train`, `y_test`, `X_train` and `X_test` by splitting the data into 75% train and 25% test datasets (2.5 marks) 
    - Set random_state to 8 and stratify the subsamples so that train and test datasets have roughly equal proportions of the target's class labels 
2. Standardise the data using `StandardScaler` library (2.5 marks)   

(Total: 5 marks) 

In [44]:
# ---- provide your code here -----
#8.1
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 8, stratify = y)

#8.2
from sklearn.preprocessing import StandardScaler
np.set_printoptions(precision=3, suppress = True) # pretty printing
sc = StandardScaler()
sc.fit(X_train)
X_train_scaled = sc.transform(X_train)
X_test_scaled = sc.transform(X_test)


---
---

### Task 4. Logistic Regression and Random Forest Classifiers and Accuracies (Total Marks: 30)


**Question 9**. 

1. Train a Logistic Regression Classifier on standardised data (5 marks)
    - Set `random_state` to 10 (don't change any other parameters)
    - Compute and print training and test dataset accuracies   
2. Train a Random Forest Classifier on standardised data (5 marks)
    - Set `random_state` to 10 (don't change any other parameters)
    - Compute and print training and test dataset accuracies

When printing accuracies round the values to three decimal places.      

(Total: 10 marks)  

In [51]:
# ---- provide your code here -----
#9.1

#training logistic model
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression(random_state=10) 
lr.fit(X_train_scaled, y_train)

#printing accuracy
print(f'Accuracy for training dataset = {lr.score(X_train_scaled, y_train):.3f}')
print(f'Accuracy for test dataset = {lr.score(X_test_scaled, y_test):.3f}')

#9.2
from sklearn.ensemble import RandomForestClassifier

forest = RandomForestClassifier(random_state=10)
forest.fit(X_train, y_train)

#printing accuracy
print(f'Accuracy for training dataset = {forest.score(X_train, y_train):.7f}')
print(f'Accuracy for test dataset = {forest.score(X_test, y_test):.7f}')

Accuracy for training dataset = 0.660
Accuracy for test dataset = 0.660
Accuracy for training dataset = 0.9771852
Accuracy for test dataset = 0.8871111


**Question 10**. 

a) Comment and compare the training and test accuracies for each classifier computed in Question 9. What can we say about the extent of overfitting for each classifier? (5 marks)   
b) Comment and compare the accuracies across the two classifiers. Which classifier provides better forecasts? (5 marks)   
c) What can you say about the presence of nonlinearities in the dataset? (10 marks)   

(Total: 20 marks)  

---- provide your written answer here ----

a)

"Overfitting" is defined when a model  captures "patterns" in the training data that do not repeat in new data. Thus, if a model has a significantly high accuracy on training dataset, following by a much smaller accuracy test dataset, we can conclude that model has a high extent for overfiting.

For logistic regression, the accuracy for training dataset (around 66%) is higher than that for test dataset (around 65.96%). This indicates that the logistic regression has achieved a nearly match accuracy when learnining the model from training dataset and use it on dataset, demonstrating a moderate overfitting level

On the other hand, for random forest classifier, the accuracy for training dataset (around 97.7%) is higher than that for test dataset (around 90%). Moreover, the surge in accuracy from training dataset to test dataset is more comparable in random forest classifier (around 97.7%-88.8%=9%) compared to logistic regression model (around 66%-65.96%=0.04%). This significant weight on training dataset accuracy (nearly perfect) demonstrate a higher overfitting rate compared to logistic regression

In general, while logistic regression has a lower learning accuracy compared to its rival, the balance between accuracy in test and training dataset makes it a less overfitting model.

b)

Comparing between the 2 classifiers, we can observe that logistic regression has a lower accuracy in test dataset (about 65%) compared to random forest classifier (around 88.8%). As a result, we can conclude that random forest classifier provide better forecast than logistic regression classifier

c)

The presence of nonlinearities in the dataset reduces the accuracy of the logistic regression classifier while random forest classifier is not affected. This is because logistic regression is a linear classifier, which computes a straight-line decision boundary to separate and classify data. However, where data points are not perfectly separable, this classifier results in low accuracy (for example a data labelled "A" is classified as "B" when it lies in the "A" area separable by decision boundary).

Conversely, the random forest classifier is capable of capturing isolated points with less extreme classification probabilities. Naturally, random forest classifier ensembles approach, which combines multiple smaller classifiers from each decision tree, results in a complex decision boundary that effectively captures the nonlinear data. Thus, it provides a more flexible decision boundary by averaging multiple decision trees, which helps in capturing the overall structure of the data, including nonlinearities. In short, the accuracy of this classifier is not related to the nonlinear property of the dataset, making it a more reliable model for prediction in this case.





---
---